### PR TITLE
support OffsetArray write

### LINKED ISSUE #46: DenseTaggedImage not generic enough for offset arrays
```julia
using ImageCore, OffsetArrays, TiffImages
img = rand(Gray, 10, 10);

TiffImages.DenseTaggedImage(img) # good

img_o = OffsetArray(img, -1, -1); 
TiffImages.DenseTaggedImage(img_0) # oops
```

### PR DESCRIPTION
**Pull request by johnnychen94**
- 021ca76: support OffsetArray write by loop over `axes` instead of hardcoded `1:length` (closes #46)
- 4eab922: `@propagate_inbounds` are added for all `getindex`/`setindex!` function. I also see two unnecessary `getindex` methods, I don't know why they exist, so you might want to double-check it.

### CODE REVIEW

**Comment by tlnagy** [comment]
Does this PR work with `AxisArrays`? I know there's the name clash between `Base.axes` and `AxisArrays.axes` that this might trigger if both are imported, i.e.

```
using AxisArrays
using TiffImages

TiffImages.load("...") # error?
```

I have a long term project to lift `OMETIFF.jl` (which relies on AxisArrays) onto `TiffImages` so I want to make sure the two are compatible.

**Comment by johnnychen94** [comment]
`AxisArray.axes` is a new function and won't override the `Base.axes` behavior, the only issue is symbol conflict, which can be easily solved by using `import` and not `using`.

For 388428b I believe there can be smarter AxisArray support, but no I don't see any issue with it yet.

File: test/runtests.jl [hunk 1]
```diff
@@ -2,6 +2,8 @@ using ColorTypes
 using ColorVectorSpace
 using Documenter
 using FixedPointNumbers
+using OffsetArrays
+import AxisArrays: AxisArray # No, we don't want to export `AxisArrays.axes`
```
Comment (tlnagy) [inline test/runtests.jl:6]: Shouldn't this say import instead of export?

**Comment by tlnagy** [comment]
Ah, yeah. The symbol conflict is exactly what I'm concerned about. I have lots of code that has `using AxisArrays` so I would prefer to avoid conflicting with that. Can we avoid the symbol error by explicitly using `Base.axes` in this PR?

**Comment by johnnychen94** [comment]
> Can we avoid the symbol error by explicitly using Base.axes in this PR?

Yes, the following also works:

```julia
using AxisArrays
using Base: axes
```

**Comment by tlnagy** [comment]
> > Can we avoid the symbol error by explicitly using Base.axes in this PR?
> 
> Yes, the following also works:
> 
> ```julia
> using AxisArrays
> using Base: axes
> ```

If you add the `using Base: axes` to this PR, I'll be happy to merge and tag a new release.

**Comment by johnnychen94** [comment]
> If you add the using Base: axes to this PR

Ha, `AxisArrays` is only used as a test dependency so it wouldn't affect the `TiffImages` functionalities at all.

Do you mean that I need to add `using Base: axes` in `src/`? I don't think we need to do so unless we're actually using `AxisArrays` inside `TiffImages`.

**Comment by tlnagy** [comment]
> Does this PR work with `AxisArrays`? I know there's the name clash between `Base.axes` and `AxisArrays.axes` that this might trigger if both are imported, i.e.
> 
> ```
> using AxisArrays
> using TiffImages
> 
> TiffImages.load("...") # error?
> ```
>
 
That's what I was asking initially. So what you're saying is that if we don't use `AxisArrays` inside of `TiffImages` then there shouldn't be any problems with the above code? Because then I'm ready to sign off on this, sorry for the confusion. 

We could also remove the AxisArrays tests since they aren't revelant.

**Comment by johnnychen94** [comment]
> So what you're saying is that if we don't use AxisArrays inside of TiffImages then there shouldn't be any problems with the above code?

Yes, there won't be any problems with the codes. When we want to use `AxisArrays` inside `TiffImages`, then we can safely avoid the symbolic conflict with `import AxisArrays` instead of `using AxisArrays`. (or with explicit `using Base: axes`).

> We could also remove the AxisArrays tests since they aren't relevant.

I'd be happier if we get more test cases (even if they're not very much relevant) 😆

**Pull request merged**